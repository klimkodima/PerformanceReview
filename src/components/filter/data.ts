export const FILTERS_NAMES = {
  TEAM: 'TEAM',
  LEVEL: 'LEVEL',
  AUDITOR: 'AUDITOR'
};

export const mockTeamFilter: string[] = ['Alpha', 'Charlie', 'X-Rays'];

export const mockLevelsFilter: string[] = [
  'Trainee',
  'Junior',
  'Middle',
  'Senior',
  'Team Lead'
];

export const mockAuditorsFilter: string[] = [
  'Alex Vagner',
  'Tom Cruise',
  'Fred Jhonson',
  'Tawes',
  'Kerry',
  'Jim',
  'Kevin',
  'Fred',
  'Franklin',
  'Ted',
  'Johnson',
  'Bob'
];
