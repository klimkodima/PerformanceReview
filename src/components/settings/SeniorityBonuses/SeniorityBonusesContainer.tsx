import { ChangeEvent, memo, ReactElement, useEffect, useState } from 'react';

import { nanoid } from 'nanoid';
import { useDispatch, useSelector } from 'react-redux';
import {
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableCell,
  TableBody,
  Paper,
  Checkbox
} from '@mui/material';
import InfoIcon from '@mui/icons-material/Info';

import {
  createNewBonusesRow,
  deleteBonusesRow,
  fetchBonusesData,
  selectBonusesData,
  SeniorityBonusType,
  updateBonusesRow
} from 'src/store/SeniorityBonuses';
import { ErrorTooltip, InfoTooltip } from 'src/components/common';
import {
  ERRORS_TEXT,
  EXPERIENCE_KEY,
  HEADER_CELLS,
  INFO_TOOLTIP
} from './constants';
import { SettingsTableBtnsBlock } from '../common';
import { handleCheckboxClick, onChangeInput } from '../common/helpers';

import { BonusesDataType } from 'src/store/SeniorityBonuses';

import '../common/SettingsTableShared.scss';
import './SeniorityBonuses.scss';

const NEW_ROW_ID = 'new-row';

const checkboxHandler =
  (id: string, selected: string[], setSelected: (arg: string[]) => void) =>
  () => {
    handleCheckboxClick(id, selected, setSelected);
  };

const onChangeHandler =
  (
    id: string,
    rowsData: BonusesDataType[],
    setBonusesData: (arg: BonusesDataType[]) => void
  ) =>
  (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    onChangeInput(event, id, rowsData, setBonusesData);
  };

type SeniorityBonusesPropsType = {
  isEditable: boolean;
};

const SeniorityBonusesContainer = memo(
  ({ isEditable }: SeniorityBonusesPropsType): ReactElement => {
    const dispatch = useDispatch();
    const rowsApi = useSelector(selectBonusesData);
    const copyBonusesData: BonusesDataType[] = JSON.parse(
      JSON.stringify(rowsApi)
    );

    const [isEdit, setIsEdit] = useState<boolean>(false);

    useEffect(() => {
      dispatch(fetchBonusesData());
    }, []);

    const [bonusesData, setBonusesData] = useState<BonusesDataType[]>(rowsApi);
    const [selected, setSelected] = useState<string[]>([]);
    const [selectedForDelete, setSelectedForDelete] =
      useState<string[]>(selected);
    const [isValid, setIsValid] = useState(false);
    const rowsData = isEdit ? bonusesData : copyBonusesData;
    const editClassName = isEdit ? 'settings-table__edit' : '';
    const { emptyField, wrongPattern, intervalError } = ERRORS_TEXT;

    useEffect(() => {
      if (selected.length !== 0) {
        setSelectedForDelete(selected);
      }
    }, [selected]);

    useEffect(() => {
      setBonusesData(copyBonusesData);
    }, [isEdit]);

    const addNewRow = () => {
      const newRow = {
        id: nanoid() + NEW_ROW_ID,
        experience: {
          value: '',
          isError: false,
          errorText: ''
        },
        bonus: {
          value: '',
          isError: false,
          errorText: ''
        }
      };

      setBonusesData([...bonusesData, newRow]);
    };

    const checkValid = (): void => {
      let errorsCounter = 0;

      const setError = (cell, errorText): void => {
        errorsCounter += 1;
        cell.isError = true;
        cell.errorText = errorText;
        setIsValid(false);
      };

      const clearError = (cell): void => {
        cell.isError = false;
        cell.errorText = '';
        setIsValid(true);
      };

      const bonusesDataCopy: BonusesDataType[] = JSON.parse(
        JSON.stringify(bonusesData)
      );

      const newRows = bonusesDataCopy.map(
        (row: BonusesDataType, index: number): BonusesDataType => {
          const noIdRow = Object.keys(row).slice(1);

          if (
            Number(bonusesDataCopy[index]?.experience?.value) <=
              Number(bonusesDataCopy[index - 1]?.experience?.value) &&
            index !== 0
          ) {
            setError(row.experience, intervalError);
          }

          for (const cell of noIdRow) {
            if (!row[cell].value) {
              setError(row[cell], emptyField);
            } else if (!/^\d+$/.test(row[cell].value as string)) {
              setError(row[cell], wrongPattern);
            } else if (errorsCounter === 0) {
              clearError(row[cell]);
            }
          }

          return row;
        }
      );
      setBonusesData(newRows);

      newRows.forEach((row: BonusesDataType): void => {
        if (
          row.id.endsWith(NEW_ROW_ID) &&
          !row.experience.isError &&
          !row.bonus.isError
        ) {
          dispatch(createNewBonusesRow(row));
        }
      });

      copyBonusesData.forEach(
        ({ experience, bonus }: BonusesDataType, index: number): void => {
          if (!newRows[index]) return;

          if (
            experience.value !== newRows[index].experience.value ||
            bonus.value !== newRows[index].bonus.value
          ) {
            const createRowData = (
              data: BonusesDataType
            ): SeniorityBonusType => ({
              id: Number(data.id),
              startInterval: Number(data.experience.value),
              bonusPercentage: Number(data.bonus.value)
            });

            if (
              !newRows[index].experience.isError &&
              !newRows[index].bonus.isError
            ) {
              dispatch(updateBonusesRow(createRowData(newRows[index])));
            }
          }
        }
      );

      if (selectedForDelete.length > 0) {
        selectedForDelete.forEach((rowId) => {
          if (!rowId.endsWith(NEW_ROW_ID)) {
            dispatch(deleteBonusesRow(rowId));
          }
        });
        setSelectedForDelete([]);
      }

      if (!errorsCounter) {
        setIsEdit(false);
        setSelected([]);
      }
    };

    return (
      <div className='settings-table seniority-bonuses'>
        <div className='settings-table__wrapper'>
          <TableContainer component={Paper} className={editClassName}>
            <Table aria-label='table'>
              <TableHead>
                <TableRow>
                  {isEdit && <TableCell />}
                  {HEADER_CELLS.map((headCell) => (
                    <TableCell key={headCell}>{headCell}</TableCell>
                  ))}
                </TableRow>
              </TableHead>
              <TableBody>
                {rowsData.map((row) => (
                  <TableRow key={row.id}>
                    {isEdit && (
                      <TableCell>
                        <Checkbox
                          className='edit__checkbox'
                          inputProps={{ 'aria-label': 'checkbox' }}
                          color='error'
                          onClick={checkboxHandler(
                            row.id,
                            selected,
                            setSelected
                          )}
                        />
                      </TableCell>
                    )}
                    {(
                      Object.keys(row) as Array<
                        keyof Omit<BonusesDataType, 'id'>
                      >
                    )
                      .slice(1)
                      .map((cell) => (
                        <TableCell key={cell + row.id}>
                          {isEdit ? (
                            <div className='seniority-bonuses__field-wrapper '>
                              <input
                                className={
                                  row[cell].isError ? 'error-input' : ''
                                }
                                type='text'
                                value={row[cell].value}
                                autoComplete='off'
                                name={cell}
                                onChange={onChangeHandler(
                                  row.id,
                                  rowsData,
                                  setBonusesData
                                )}
                              />
                              {cell === EXPERIENCE_KEY && (
                                <InfoTooltip
                                  title={INFO_TOOLTIP}
                                  icon={
                                    <InfoIcon
                                      className='info-icon'
                                      color='primary'
                                    />
                                  }
                                />
                              )}
                              {row[cell].isError && (
                                <div className='error-icon'>
                                  <ErrorTooltip title={row[cell].errorText} />
                                </div>
                              )}
                            </div>
                          ) : (
                            row[cell].value
                          )}
                        </TableCell>
                      ))}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
          {isEditable && (
            <SettingsTableBtnsBlock
              isEdit={isEdit}
              isValid={isValid}
              setIsEdit={setIsEdit}
              addNewRow={addNewRow}
              setSelected={setSelected}
              rowsState={bonusesData}
              selected={selected}
              setRowsState={setBonusesData}
              rowsApi={copyBonusesData}
              checkValidation={checkValid}
            />
          )}
        </div>
      </div>
    );
  }
);

export default SeniorityBonusesContainer;

SeniorityBonusesContainer.displayName = 'SeniorityBonusesContainer';
