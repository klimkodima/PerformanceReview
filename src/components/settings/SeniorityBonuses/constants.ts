import { t } from 'i18next';

export const HEADER_CELLS: string[] = [
  t('seniority_bonuses.header_cell_experience'),
  t('seniority_bonuses.header_cell_bonus')
];

export const CELL_NAMES = {
  experience: t('seniority_bonuses.experience'),
  bonus: t('seniority_bonuses.bonus')
};

export const ERRORS_TEXT = {
  wrongPattern: t('seniority_bonuses.wrong_pattern'),
  emptyField: t('seniority_bonuses.empty_field'),
  intervalError: t('seniority_bonuses.interval_error')
};

export const INFO_TOOLTIP = t('seniority_bonuses.info_tooltip_text');

export const EXPERIENCE_KEY = 'experience';
