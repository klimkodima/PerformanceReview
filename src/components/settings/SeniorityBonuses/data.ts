import { nanoid } from 'nanoid';

import { BonusesDataType } from 'src/store/SeniorityBonuses';

export const mockBonusesData: BonusesDataType[] = [
  {
    id: nanoid(),
    experience: {
      value: '12',
      isError: false,
      errorText: ''
    },
    bonus: {
      value: '10',
      isError: false,
      errorText: ''
    }
  },
  {
    id: nanoid(),
    experience: {
      value: '15',
      isError: false,
      errorText: ''
    },
    bonus: {
      value: '25',
      isError: false,
      errorText: ''
    }
  },
  {
    id: nanoid(),
    experience: {
      value: '25',
      isError: false,
      errorText: ''
    },
    bonus: {
      value: '30',
      isError: false,
      errorText: ''
    }
  },
  {
    id: nanoid(),
    experience: {
      value: '30',
      isError: false,
      errorText: ''
    },
    bonus: {
      value: '35',
      isError: false,
      errorText: ''
    }
  }
];
