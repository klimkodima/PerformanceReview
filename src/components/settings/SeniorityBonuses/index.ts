export { default as SeniorityBonusesEdit } from './SeniorityBonusesEdit';
export { default as SeniorityBonuses } from './SeniorityBonuses';
