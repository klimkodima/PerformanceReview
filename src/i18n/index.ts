import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import common from './locales.en.json';

const resources = {
  en: {
    translation: common
  }
};

const setI18nextConfig = async () =>
  await i18n.use(LanguageDetector).use(initReactI18next).init({
    resources,
    fallbackLng: 'en'
  });

void setI18nextConfig();

export default i18n;
