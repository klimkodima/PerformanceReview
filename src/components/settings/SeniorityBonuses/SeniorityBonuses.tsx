import { ReactElement } from 'react';

import SeniorityBonusesContainer from './SeniorityBonusesContainer';

const SeniorityBonuses = (): ReactElement => (
  <SeniorityBonusesContainer isEditable={false} />
);

export default SeniorityBonuses;
