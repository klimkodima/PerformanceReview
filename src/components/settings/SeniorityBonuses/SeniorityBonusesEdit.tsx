import React, { ReactElement } from 'react';

import SeniorityBonusesContainer from './SeniorityBonusesContainer';

const SeniorityBonusesEdit = (): ReactElement => {
  return <SeniorityBonusesContainer isEditable={true} />;
};

export default SeniorityBonusesEdit;
