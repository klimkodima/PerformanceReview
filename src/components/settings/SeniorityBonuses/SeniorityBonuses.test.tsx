import { I18nextProvider } from 'react-i18next';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import store from 'src/store';
import i18n from 'src/i18n';
import { setBonusesData } from 'src/store/SeniorityBonuses';
import { SeniorityBonusesEdit } from './index';
import { mockBonusesData } from './data';
import SeniorityBonusesContainer from './SeniorityBonusesContainer';

beforeEach(() => {
  store.dispatch(setBonusesData(mockBonusesData));
});

describe('Seniority bonuses container tests', () => {
  test('Seniority bonuses snapshot', () => {
    const component = renderer
      .create(
        <I18nextProvider i18n={i18n}>
          <Provider store={store}>
            <SeniorityBonusesEdit />
          </Provider>
        </I18nextProvider>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Seniority bonuses component rendered', () => {
    const { getByRole, getAllByRole } = render(
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <SeniorityBonusesEdit />
        </Provider>
      </I18nextProvider>
    );

    expect(getByRole('table')).toBeInTheDocument();
    expect(getAllByRole('columnheader')).toHaveLength(2);
    expect(getAllByRole('row')).toHaveLength(5);
    expect(getAllByRole('cell')).toHaveLength(8);
  });
});

describe('Seniority bonuses container tests without edit button', () => {
  test('Seniority bonuses container component rendered', () => {
    const { getByRole, getAllByRole, queryByRole } = render(
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <SeniorityBonusesContainer isEditable={false} />
        </Provider>
      </I18nextProvider>
    );

    expect(getByRole('table')).toBeInTheDocument();
    expect(getAllByRole('columnheader')).toHaveLength(2);
    expect(getAllByRole('row')).toHaveLength(5);
    expect(getAllByRole('cell')).toHaveLength(8);
    expect(queryByRole('button')).toBeNull();
  });
});

describe('Seniority bonuses table edit button and checkbox tests', () => {
  test('Seniority bonuses table edit button and checkbox click test', () => {
    const { getByRole, getAllByRole, queryAllByRole } = render(
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <SeniorityBonusesEdit />
        </Provider>
      </I18nextProvider>
    );

    expect(queryAllByRole('checkbox')).toHaveLength(0);
    userEvent.click(getByRole('button', { name: 'Edit Table' }));

    const checkboxes = getAllByRole('checkbox');
    expect(checkboxes).toHaveLength(4);
    expect(getAllByRole('textbox')).toHaveLength(8);

    userEvent.click(checkboxes[0]);
    expect(checkboxes[0]).toBeChecked();
  });

  test('must change row', () => {
    const { getAllByRole } = render(
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <SeniorityBonusesContainer isEditable={true} />
        </Provider>
      </I18nextProvider>
    );

    const editBtn = screen.getByRole('button', { name: 'Edit Table' });
    userEvent.click(editBtn);

    const textBoxes = getAllByRole('textbox');

    expect(textBoxes[0]).toHaveValue('12');
    userEvent.clear(textBoxes[0]);
    userEvent.type(textBoxes[0], '1');
    expect(textBoxes[0]).toHaveValue('1');

    expect(textBoxes[1]).toHaveValue('10');
    userEvent.clear(textBoxes[1]);
    userEvent.type(textBoxes[1], '5');
    expect(textBoxes[1]).toHaveValue('5');
  });
});
