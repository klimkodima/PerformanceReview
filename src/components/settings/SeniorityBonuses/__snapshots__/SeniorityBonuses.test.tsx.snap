// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seniority bonuses container tests Seniority bonuses snapshot 1`] = `
<div
  className="settings-table seniority-bonuses"
>
  <div
    className="settings-table__wrapper"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="table"
        className="MuiTable-root css-rqglhn-MuiTable-root"
        role={null}
      >
        <thead
          className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Experience (months)
            </th>
            <th
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Bonus
            </th>
          </tr>
        </thead>
        <tbody
          className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          role={null}
        >
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              12
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              10
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              15
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              25
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              25
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              30
            </td>
          </tr>
          <tr
            className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
            role={null}
          >
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              30
            </td>
            <td
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              35
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="settings-table__btns-block"
    >
      <button
        className="MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root btns-block__edit-btn css-sxix9q-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </span>
        Edit Table
      </button>
    </div>
  </div>
</div>
`;
